### ISSUE TITLE
Initial Intel Trust Domain Extensions (TDX) Enablement

### ISSUE THREAD

**Issue by sameo** [cloud-hypervisor/cloud-hypervisor#2226]
[Intel Trust Domain Extensions](https://software.intel.com/content/dam/develop/external/us/en/documents/tdx-whitepaper-v4.pdf) are a new x86 ISA extension designed for running Virtual Machines (VMs) inside hardware isolated Trust Domains (TDs).
With TDX, any non -TD software (platform firmware, host kernel, hypervisor, VMM, etc) is removed from the VM Trusted Computing Base (TCB), allowing for protecting guest workloads from the platform or cloud owners/operators.

This issue describes a proposal for implementing the initial TDX support in Cloud Hypervisor.

- [TDX 101](#tdx-101)
- [Motivation](#motivation)
- [Scope](#scope)
- [TDX Software Stack](#tdx-software-stack)
  * [Trust Domain Virtual Firmware (TDVF)](#trust-domain-virtual-firmware--tdvf-)
  * [Host Linux Kernel and KVM](#host-linux-kernel-and-kvm)
    + [`KVM_TDX_CAPABILITIES`](#-kvm-tdx-capabilities-)
    + [`KVM_TDX_INIT_VM`](#-kvm-tdx-init-vm-)
    + [`KVM_TDX_INIT_CPU`](#-kvm-tdx-init-cpu-)
    + [`KVM_TDX_INIT_MEM_REGION`](#-kvm-tdx-init-mem-region-)
    + [`KVM_TDX_FINALIZE_VM`](#-kvm-tdx-finalize-vm-)
  * [Guest Linux Kernel](#guest-linux-kernel)
- [Implementation](#implementation)
  * [Requirements and Steps](#requirements-and-steps)
  * [TDVF support](#tdvf-support)
    + [TDVF parsing](#tdvf-parsing)
    + [TD HOB](#td-hob)
  * [TD vCPUs initial state](#td-vcpus-initial-state)
- [Links](#links)

# TDX 101

In order to enforce the proposed security guarantees, TDX defines a new CPU mode called Secure Arbitration Mode (SEAM).
In that mode, the CPU will run an Intel provided and signed software module, the Intel-TDX module, which acts as a trusted intermediary between the TD guest and the host.
It is installed and loaded by an Authenticated Code Module (ACM): The SEAM Loader (SEAMLDR). The VMM is responsible for invoking the SEAMLDR ACM, which will then have the SEAMLDR load the Intel TDX-module and record its measurement into TDX hardware-measurement registers.

Any interaction between the host and the TD is handled by the Intel TDX-module, and the latter provides an interface for the VMM to handle the TD lifecycle, without it being able to intercept
TD accesses to TD assigned resources (CRs, MSRs, TSC, etc). The main interface between the VMM and the TD are the `SEAMCALL` and `SEAMRET` instructions. The former places the CPU in SEAM-VMX-root operations (equivalent to VMENTER) while the latter gives control back to the VMM (a la VMEXIT).

To provide guests with that level of isolation, TDX can provide 2 major capabilities:

1. Memory and CPU state confidentiality and integrity. Both the guest memory and its CPUs state (TD state, VMCS, etc) are encrypted and integrity protected.
2. Remote attestation. The guest owner can verify that the guest is running on an Intel TDX-enabled platform, that it's running inside a TD, and that it's executing the expected and measured software stack, before e.g. provisioning it with confidential data over a secured communication channel.

Please read the [Intel TDX Whitepaper](https://software.intel.com/content/dam/develop/external/us/en/documents/tdx-whitepaper-final9-17.pdf), the [Intel TDX Architecture Specifications](https://software.intel.com/content/dam/develop/external/us/en/documents/intel-tdx-cpu-architectural-specification.pdf) and the [Intel TDX-module Architecture Specifications](https://software.intel.com/content/dam/develop/external/us/en/documents/intel-tdx-cpu-architectural-specification.pdf) for (much) more details on the TDX internals.

# Motivation

As a cloud focused VMM, it makes a lot of sense for Cloud Hypervisor to support features aiming at providing cloud tenants with the latest security and confidentiality features, like e.g. Intel TDX.

# Scope

The initial scope for this proposal is restricted to the memory and CPU state encryption and integrity features. Remote attestation is left out of this initial enablement proposal for now.

Remote attestation is mostly a protocol handled between one of the TD software stack component (Firmware or userspace attestation agent), and a remote verifier, and the VMM plays only a little role in that path.
As such this scope restriction is minimal. The bulk of a VMM role with regard to TDX is about interacting with the TD firmware and enabling the high level memory and CPU state encryption capabilities, through KVM.

# TDX Software Stack

In order to understand the different steps needed for a VMM to boot and manage a TD guest, we need first to understand the different layers composing a TD guest software stack.

## Trust Domain Virtual Firmware (TDVF)

Keeping the host platform out of the TD's trust boundary means the VMM can no longer directly set a vCPU RIP or more generally load and directly boot a guest kernel.
TD vCPUs boot at a fixed address, in protected mode with paging disabled. It's then up to a virtual firmware, running in the TD, to set things up (long mode transition,
initial page table setup, etc) to e.g. a guest kernel boot.

Because TDX follows a specific vCPUs initialization procedure, this virtual firmware is TDX specific and naturally called Trust Domain Virtual Firmware (TDVF).
The initial TDVF reference implementation is based on the TianoCore implementation and available at https://github.com/tianocore/edk2-staging/tree/TDVF

Although the VMM is kept out of the guest's TCB, it must still be able to describe the resources (CPUs, memory, MMIO space) the Intel TDX-module should allocate
to the TD guest. The VMM builds that description through a TD Hand Off Block (TD-HOB), containing multiple resource descriptors, that is passed to the TDVF.
The TDVF will parse and measure the TD-HOB, and then translate it into a DXE HOB before jumping to the DXE module. 

See the [TDVF Design Guide](https://software.intel.com/content/dam/develop/external/us/en/documents/tdx-virtual-firmware-design-guide-rev-1.pdf) for more details.

## Host Linux Kernel and KVM

The initial Linux kernel support for TDX was submitted on [2020/11/16](https://lkml.org/lkml/2020/11/16/1106).
The latest version of the TDX host patchset is available [here](https://github.com/intel/tdx/tree/kvm).

From a VMM enablement perspective, the most relevant part of that submission is the KVM API extension for TDX.
Following AMD's SEV API, the TDX KVM API is multiplexed through the VM level `KVM_MEMORY_ENCRYPT_OP` KVM ioctl.

There are 5 TDX `KVM_MEMORY_ENCRYPT_OP` sub-ioctls, all of them take a `kvm_tdx_cmd` argument:

```
struct kvm_tdx_cmd {
    __u32 id;             /* tdx_cmd_id */
    __u32 metadata;       /* sub command specific */
    __u64 data;           /* sub command specific */
  };
```

### `KVM_TDX_CAPABILITIES`

This ioctl returns a subset of the `TDSYSINFO_STRUCT` fetched by the `TDH.SYS.INFO` TDX SEAM call, packed into a `kvm_tdx_capabilities` structure:

```C
struct kvm_tdx_capabilities {
	__u64 attrs_fixed0;
	__u64 attrs_fixed1;
	__u64 xfam_fixed0;
	__u64 xfam_fixed1;

	__u32 nr_cpuid_configs;
	struct kvm_tdx_cpuid_config cpuid_configs[0];
};
```

In practice, this ioctl allows for adjusting the supported CPUID set.

### `KVM_TDX_INIT_VM`

This ioctl does the TDX initialization for a VM, and is essentially a wrapper around the `TDH.MNG.INIT` TDX SEAM call.

It takes a pointer to a `kvm_tdx_init_vm` as an argument, through the TDX command `data` field:

```C
struct kvm_tdx_init_vm {
    __u32 max_vcpus;
    __u32 reserved;
    __u64 attributes;
    __u64 cpuid;  /* pointer to struct kvm_cpuid2 */
  };
```

### `KVM_TDX_INIT_CPU`

This ioctl does additional, TDX-specific vCPU initialization and is essentially a wrapper around the `TDH.VP.INIT` TDX SEAM call.
Its only argument is the initial value of the guest TD vCPU RCX register.

### `KVM_TDX_INIT_MEM_REGION`

This ioctl encrypts a memory regions. It calls the `TDH.MEM.PAGE.ADD` TDX SEAM call.

Each memory region to be encrypted is described through the following structure:

```C
struct kvm_tdx_init_mem_region {
	__u64 source_addr;
	__u64 gpa;
	__u64 nr_pages;
};
```

passed as a pointer, through the TDX command `data` field.

If the TDX command `metadata` field is set to `KVM_TDX_MEASURE_MEMORY_REGION`, then the TDX-module will extend the TD measurement with that
memory region. This is equivalent to calling the `TDH.MR.EXTEND` TDX SEAM call.

### `KVM_TDX_FINALIZE_VM`

This ioctl takes no arguments, completes the initial TD measurement, and marks it ready to be running.
This is a wrapper around the `TDH.MR.FINALIZE` TDX SEAM call.

## Guest Linux Kernel

The latest guest Linux kernel support for TDX is available [here](https://github.com/intel/tdx/tree/guest).

In order to run as a TD guest, the Linux kernel must support the following features:

* TDX IO
  * Port IO is handled from the TD guest, by unrolling IO port `#VE` exceptions into TDVMCALLs. That will eventually translate into an unmodified VMM PIO exit handlers.
  * Virtio: Essentially, the TD guest kernel converts the private (TDX encrypted) virtio queue pages into shared ones in order to seamlessly use the guest virtio drivers and the VMM virtio compliant device implementations. 
  * Direct device assignment: DMA between MMIO device buffers and the TD guest must happen in the TD shared memory address space. Converting DMA buffers to shared memory is handled by the TDX guest kernel.
* ACPI SKVL (Storage Volume Key Label) is an ACPI table for passing storage encryption keys from the TDVF to the guest kernel. Again, this should not have any impact on the VMM enablement path.
* Remote attestation driver. This is a kernel interface for the guest to trigger the attestation process by requesting a TD quote from the TDX-module. Attestation is out of scope for our initial enablement effort.

Overall, in the context of our initial TDX enablement effort with Cloud Hypervisor, the TDX specific guest Linux kernel changes should be transparent to the VMM implementation.

# Implementation

## Requirements and Steps

In addition to its legacy VM (Non TDX) support, a KVM based VMM aiming at booting and managing TD guests should be able to build a TD and initialize it:

* [x] Parse a TDVF
* [x] Load a TDVF at the guest reset vector (0xFFFFFFF0)
* [x] Create a TD HOB describing the guest resources
* [x] Add the TD HOB to the loaded TDVF metadata section
* [x] Create a VM and initialize it as a TD (`KVM_TDX_INIT_VM`)
* [x] Create vCPUs and initialize them as TD vCPUs (`KVM_TDX_INIT_CPU`)
* [x] Encrypt all memory regions described by the TDVF entries (`KVM_TDX_INIT_MEM_REGION`)
* [x] Finalize the TD measurements before starting the VM (`KVM_TDX_FINALIZE_VM`)

## TDVF support

A large chunk of the implementation steps will be about supporting the OVMF based TDVF, by parsing, loading, and adding the TD HOB to it.
The [TDVF Design Guide](https://software.intel.com/content/dam/develop/external/us/en/documents/tdx-virtual-firmware-design-guide-rev-1.pdf) provides the complete specifications to follow for implementing the features below: 

### TDVF parsing

The VMM must parse the TDVF's metadata as it describes all the firmware boot code (BFV), configuration (CVF), temporary firmware memory and TD memory to be initially encrypted and measured by the TDX-module.
The TDVF metadata is placed in in the TDVF descriptor, which is made of multiple TDVF sections.
The VMM also must create a TD HOB describing the VMM resources and have the TDVF descriptor point at it

### TD HOB

The TD HOB is built and loaded by the VMM.
It is actually a list of HOBs, starting from a PHIT HOB and followed by resource HOBs describing the physical memory and the MMIO regions.

The physical memory HOBs in the TD HOB list contain one resource descriptor per entry in the VMM generated `e820` table for the TD guest.

## TD vCPUs initial state

All vCPUs in a TD will start from a fixed, architecurally defined state. The VMM KVM vCPU settings will be overwritten by the TD vCPU architectural state.

This initial TD vCPU state is very close to the CPU architectural state after `INIT`, except that the vCPU will be initialized in protected mode (not real mode) with paging disabled.
Again, the VMM initial page tables, GPRs, CRs or SRs settings will have no effect on the actual TD vCPUs initial state.

See chapter 8 of the [Intel TDX-module Architecture Specifications](https://software.intel.com/content/dam/develop/external/us/en/documents/intel-tdx-cpu-architectural-specification.pdf) for more details.

# Links

* [Intel TDX Whitepaper](https://software.intel.com/content/dam/develop/external/us/en/documents/tdx-whitepaper-final9-17.pdf)
* [Intel TDX Architecture Specifications](https://software.intel.com/content/dam/develop/external/us/en/documents/intel-tdx-cpu-architectural-specification.pdf)
* [Intel TDX-module Architecture Specifications](https://software.intel.com/content/dam/develop/external/us/en/documents/intel-tdx-cpu-architectural-specification.pdf)
* [Intel TDX Guest Host Communication Interface (GHCI)](https://software.intel.com/content/dam/develop/external/us/en/documents/intel-tdx-guest-hypervisor-communication-interface.pdf)
* [TDVF Design Guide](https://software.intel.com/content/dam/develop/external/us/en/documents/tdx-virtual-firmware-design-guide-rev-1.pdf)

**Comment by sameo** [comment]
cc @rbradford @sboeuf @arronwy @Jiangbo-W

**Comment by rbradford** [comment]
I believe all TDX work is complete.
